BParkingNano/plugins/TriggeringMuonProducer.cc

```diff
@@ -0,0 +1,173 @@
+
+using namespace std;
+
+
+float MuonMass_ = 0.10565837;
+
+class TriggeringMuonProducer : public edm::EDProducer {
+    
+public:
+    
+    explicit TriggeringMuonProducer(const edm::ParameterSet &iConfig);
+    
+    ~TriggeringMuonProducer() override {};
+    
+    
+private:
+
+    virtual void produce(edm::Event&, const edm::EventSetup&);
+
+
+    edm::EDGetTokenT<std::vector<pat::Muon>> muonSrc_;
+    edm::EDGetTokenT<edm::TriggerResults> triggerBits_;
+    edm::EDGetTokenT<std::vector<pat::TriggerObjectStandAlone>> triggerObjects_;
+    edm::EDGetTokenT<pat::PackedTriggerPrescales> triggerPrescales_;
+    edm::EDGetTokenT<reco::VertexCollection> vertexSrc_;
+
+    bool debug;
+};
+
+
+TriggeringMuonProducer::TriggeringMuonProducer(const edm::ParameterSet &iConfig):
+  muonSrc_( consumes<std::vector<pat::Muon>> ( iConfig.getParameter<edm::InputTag>( "muonCollection" ) ) ),
+  triggerBits_(consumes<edm::TriggerResults>(iConfig.getParameter<edm::InputTag>("bits"))),
+  triggerObjects_(consumes<std::vector<pat::TriggerObjectStandAlone>>(iConfig.getParameter<edm::InputTag>("objects"))),
+  triggerPrescales_(consumes<pat::PackedTriggerPrescales>(iConfig.getParameter<edm::InputTag>("prescales"))),
+  vertexSrc_( consumes<reco::VertexCollection> ( iConfig.getParameter<edm::InputTag>( "vertexCollection" ) ) )
+{
+    produces<pat::CompositeCandidateCollection>();
+    debug = false;
+}
+
+
+
+void TriggeringMuonProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+    
+
+    edm::Handle<reco::VertexCollection> vertexHandle;
+    iEvent.getByToken(vertexSrc_, vertexHandle);
+    const reco::Vertex & PV = vertexHandle->front();
+
+    if(debug) std::cout << " BToKstllProducer::identifyTriggeringMuons " << std::endl;
+
+    edm::Handle<edm::TriggerResults> triggerBits;
+    iEvent.getByToken(triggerBits_, triggerBits);
+    const edm::TriggerNames &names = iEvent.triggerNames(*triggerBits);
+
+    std::vector<std::vector<float>> triggeringMuons;
+    //taken from https://twiki.cern.ch/twiki/bin/view/CMSPublic/WorkBookMiniAOD2016#Trigger
+    edm::Handle<std::vector<pat::TriggerObjectStandAlone>> triggerObjects;
+    iEvent.getByToken(triggerObjects_, triggerObjects);
+    if(debug) std::cout << "\n TRIGGER OBJECTS " << std::endl;
+
+    for (pat::TriggerObjectStandAlone obj : *triggerObjects) { // note: not "const &" since we want to call unpackPathNames
+      obj.unpackFilterLabels(iEvent, *triggerBits);
+      obj.unpackPathNames(names);
+
+      bool isTriggerMuon = false;
+      for (unsigned h = 0; h < obj.filterIds().size(); ++h)
+	if(obj.filterIds()[h] == 83){ 
+	  isTriggerMuon = true; 
+	  if(debug) std::cout << "\t   Type IDs:   " << 83;  //83 = muon
+	  break;
+	} 
+
+      if(!isTriggerMuon) continue; 
+      for (unsigned h = 0; h < obj.filterLabels().size(); ++h){
+	std::string filterName = obj.filterLabels()[h];
+	if(filterName.find("hltL3") != std::string::npos  && filterName.find("Park") != std::string::npos){
```

Comment:
This can be pretty slow, it would be better to provide the list of filters to fully match against and store the strings into an `std::unordered_set` and just look if the filter label is in the set.

Reply:
It seems quite efficient at the moment, but sure can be improved. I mark as lower priority for now and will update.